ui/components/src/ThresholdsEditor/ThresholdsEditor.tsx

```diff
@@ -0,0 +1,238 @@
+import DeleteIcon from 'mdi-material-ui/DeleteOutline';
+import PlusIcon from 'mdi-material-ui/Plus';
+import { Stack } from '@mui/system';
+import { ThresholdOptions } from '@perses-dev/core';
+import { InfoTooltip } from '../InfoTooltip';
+import { useChartsTheme } from '../context/ChartsThemeProvider';
+import { OptionsEditorGroup } from '../OptionsEditorLayout';
+import { ThresholdColorPicker } from './ThresholdColorPicker';
+
+interface ThresholdsEditorProps {
+  thresholds?: ThresholdOptions;
+  onChange: (thresholds: ThresholdOptions) => void;
+}
+
+const DEFAULT_STEP = 10;
+
+export function ThresholdsEditor({ thresholds, onChange }: ThresholdsEditorProps) {
+  const {
+    thresholds: { defaultColor, palette },
+  } = useChartsTheme();
+  const defaultThresholdColor = thresholds?.default_color ?? defaultColor;
+
+  const [steps, setSteps] = useState(thresholds?.steps);
+  useEffect(() => {
+    setSteps(thresholds?.steps);
+  }, [thresholds?.steps]);
+
+  // every time a new threshold is added, we want to focus the recently added input
+  const recentlyAddedInputRef = useRef<HTMLInputElement | null>(null);
+  const focusRef = useRef(false);
+  useEffect(() => {
+    if (!recentlyAddedInputRef.current || !focusRef.current) return;
+    recentlyAddedInputRef.current?.focus();
+    focusRef.current = false;
+  }, [steps?.length]);
+
+  const handleThresholdValueChange = (e: React.ChangeEvent<HTMLInputElement>, i: number) => {
+    setSteps(
+      produce(steps, (draft) => {
+        const step = draft?.[i];
+        if (step) {
+          step.value = Number(e.target.value);
+        }
+      })
+    );
+  };
+
+  const handleThresholdColorChange = (color: string, i: number) => {
+    if (thresholds !== undefined) {
+      onChange(
+        produce(thresholds, (draft) => {
+          if (draft.steps !== undefined) {
+            const step = draft.steps[i];
+            if (step) {
+              step.color = color;
+            }
+          }
+        })
+      );
+    }
+  };
+
+  const handleDefaultColorChange = (color: string) => {
+    if (thresholds !== undefined) {
+      onChange(
+        produce(thresholds, (draft) => {
+          draft.default_color = color;
+        })
+      );
+    }
+  };
+
+  // sort thresholds in ascending order every time an input blurs
+  const handleThresholdBlur = () => {
+    if (steps !== undefined) {
+      const sortedSteps = [...steps];
+      sortedSteps.sort((a, b) => a.value - b.value);
+      if (thresholds !== undefined) {
+        onChange(
+          produce(thresholds, (draft) => {
+            draft.steps = sortedSteps;
+          })
+        );
+      }
+    }
+  };
+
+  const deleteThreshold = (i: number): void => {
+    if (thresholds !== undefined) {
+      const updatedThresholds = produce(thresholds, (draft) => {
+        if (draft.steps) {
+          draft.steps.splice(i, 1);
+        }
+      });
+      onChange(updatedThresholds);
+    }
+  };
+
```

Comment:
Optional suggestion -- You've got a decent amount of logic around adjusting these values that's probably a bit of a pain to test in the context of a react component. If you pulled the core logic out into a few utility functions that you call from the component, you could more easily write some unit tests on those functions outside the react context. That said, you've already written a bunch of tests for the components, so may not be worth refactoring now.

Reply:
oh nice i'll keep that in mind for next time, and writing those tests in react context actually helped me uncover a few accessibility issues